Make proxyv2 image building operations universal

GLYASAI opened the pull request:
ref: #24457

When building a proxyv2 image, you can specify the name of Sidecar through environment variables, so that various data planes can be easily integrated into Istio

1. Use environment variable SIDECAR to replace hard-coded envoy
2. Download WebAssembly only when sidecar is envoy

By default, use envoy to build, the command is as follows:

```
make docker.proxyv2
```

If you use other data planes, such as mosn, use the following command to build

```
ISTIO_ENVOY_VERSION=v0.13.0 ISTIO_ENVOY_RELEASE_URL=https://github.com/GLYASAI/mosn/releases/download/0.13.0/mosn.tgz ISTIO_ENVOY_MACOS_RELEASE_URL=https://github.com/GLYASAI/mosn/releases/download/0.13.0/mosn-macos.tar.gz ISTIO_ENVOY_MACOS_RELEASE_NAME=mosn-0.13.0 SIDECAR=mosn make docker.proxyv2
```

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[X] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

istio-testing commented:
@GLYASAI: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

istio-testing commented:
Hi @GLYASAI. Thanks for your PR.

I'm waiting for a [istio](https://github.com/orgs/istio/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/istio/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=istio%2Fistio).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

GLYASAI commented:
@googlebot I consent

wangfakang commented:
@howardjohn  Plz review,thx.

rootsongjc commented:
@googlebot I consent.

GLYASAI commented:
@googlebot I consent.

rootsongjc commented:
/ok-to-test

istio-testing commented:
@GLYASAI: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | 93426e4a61f4e81ea5c511f71661290428d7637e | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/25146/lint_istio/15149) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

wangfakang commented:
@googlebot I consent.

istio-testing commented:
@GLYASAI: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | 93426e4a61f4e81ea5c511f71661290428d7637e | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/25146/lint_istio/15149) | `/test lint_istio`
integ-telemetry-k8s-tests_istio | 93426e4a61f4e81ea5c511f71661290428d7637e | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/25146/integ-telemetry-k8s-tests_istio/15144) | `/test integ-telemetry-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn commented:
How many other customizations will MOSN need in the future? It seems to me like it may be more practical for MOSN to have its own dockerfile?

howardjohn commented:
Not strongly against this right now, just worried about the future for both projects - I don't want MOSN to be slowed down by Istio and also don't want Istio to take on too much complexity (in the future - this PR is simple).

I am also slightly worried about the variable rename, as I know a number of build pipelines setting these currently.

wangfakang commented:
> Not strongly against this right now, just worried about the future for both projects - I don't want MOSN to be slowed down by Istio and also don't want Istio to take on too much complexity (in the future - this PR is simple).
> 
> I am also slightly worried about the variable rename, as I know a number of build pipelines setting these currently.

@howardjohn Thank's your reply. This PR is used to decouple Istio from specific data plane when building `proxyv2` images，for example, it can be via parameter specifies the name of the sidecar, which is convenient for compiling other data planes(MOSN is only one of them). On the other hand, the default is to use the Envoy data plane, which is compatible with the previous version.

wangfakang commented:
> How many other customizations will MOSN need in the future? It seems to me like it may be more practical for MOSN to have its own dockerfile?

@howardjohn There will be no special customization, and we‘ll follow the Istio standard. As mentioned above, this PR is designed to make it easy for istio's consumers to quickly integrate other data planes（default use Envoy） into Istio.

champly commented:
> ref: #24457
> 
> When building a proxyv2 image, you can specify the name of Sidecar through environment variables, so that various data planes can be easily integrated into Istio
> 
> 1. Use environment variable SIDECAR to replace hard-coded envoy
> 2. Download WebAssembly only when sidecar is envoy
> 
> By default, use envoy to build, the command is as follows:
> 
> ```
> make docker.proxyv2
> ```
> 
> If you use other data planes, such as mosn, use the following command to build
> 
> ```
> ISTIO_ENVOY_VERSION=v0.13.0 ISTIO_ENVOY_RELEASE_URL=https://github.com/GLYASAI/mosn/releases/download/0.13.0/mosn.tgz ISTIO_ENVOY_MACOS_RELEASE_URL=https://github.com/GLYASAI/mosn/releases/download/0.13.0/mosn-macos.tar.gz ISTIO_ENVOY_MACOS_RELEASE_NAME=mosn-0.13.0 SIDECAR=mosn make docker.proxyv2
> ```
> 
> [ ] Configuration Infrastructure
> [ ] Docs
> [ ] Installation
> [ ] Networking
> [ ] Performance and Scalability
> [ ] Policies and Telemetry
> [ ] Security
> [X] Test and Release
> [ ] User Experience
> [ ] Developer Infrastructure

example should use `SIDECAR` replace `ENVOY`, such as

```shell
ISTIO_SIDECAR_VERSION=v0.13.0 ISTIO_SIDECAR_RELEASE_URL=https://github.com/GLYASAI/mosn/releases/download/0.13.0/mosn.tgz ISTIO_SIDECAR_MACOS_RELEASE_URL=https://github.com/GLYASAI/mosn/releases/download/0.13.0/mosn-macos.tar.gz ISTIO_SIDECAR_MACOS_RELEASE_NAME=mosn-0.13.0 SIDECAR=mosn make docker.proxyv2
```

rootsongjc commented:
/assign howardjohn

wangfakang commented:
/assign @howardjohn

howardjohn commented:
Do we need to change all the variable names? MOSN can still override the variables even if it's named "envoy". There are so many CI pipelines using these variables that will be painful to update for no real reason.

Also is it so bad to have the wasm bundles added? They are quite small... If you want to customize the image more I highly encourage building your own images

howardjohn commented on pilot/docker/Dockerfile.proxyv2:
```diff
@@ -47,8 +48,7 @@ COPY pilot-agent /usr/local/bin/pilot-agent
 
-COPY stats-filter.wasm /etc/istio/extensions/stats-filter.wasm
-COPY metadata-exchange-filter.wasm /etc/istio/extensions/metadata-exchange-filter.wasm
+COPY . *.wasm /etc/istio/extensions/
```
isn't this copying `.` which is everything?

GLYASAI replied:
MOSN does not have two files, stats-filter.wasm and metadata-exchange-filter.wasm. COPY. *.Wasm /etc/istio/extensions/ is to prevent Dockerfile from reporting errors when there are no two files

howardjohn replied:
You are copying the entire docker context to the extensions folder though rather than nothing, I think.

GLYASAI replied:
problem solved.

GLYASAI commented:
> Do we need to change all the variable names? MOSN can still override the variables even if it's named "envoy". There are so many CI pipelines using these variables that will be painful to update for no real reason.

Now, after changing the variable names, the envoy variables is still supported.

GLYASAI commented:
> I don't think renaming this benefits anyone, and only causes additional confusion and complexity. If you do not want to name a variable "envoy" (which has absolutely no impact on users of MOSN) then you should create your own dockerfile hosted in the MOSN org or istio-ecosystem.
> 
> Attempting to rename all mentions of "envoy" to a generic term does not help anyone and just causes wasted time and confusion in my opinion.

The purpose of this is to make istio more standardized and more convenient for istio users.

If you think that renaming variables will waste time and conclusion, we can still use the envoy variables.

GLYASAI commented:
@howardjohn I have changed the variable back. Plz review, thx.

GLYASAI commented:
I got confused now, what is 'we have replaced envoy with SIDECAR' mean?

GLYASAI commented:
> I don't get the purpose of this now - we have replaced envoy with SIDECAR but the behavior I'd identical?

I got confused now, what is 'we have replaced envoy with SIDECAR' mean?

GLYASAI commented:
@howardjohn Could you please explain it in more detail? thx.

wangfakang commented:
> I don't get the purpose of this now - we have replaced envoy with SIDECAR but the behavior I'd identical?

@howardjohn  Now the build script for the `proxyv2` image has some logic hardcoded, So this PR is designed to make it easy for istio's user to quickly integrate other data planes（default use Envoy) into Istio via some istio's environment variables.
